lib/appsignal/oban.ex

```diff
@@ -0,0 +1,172 @@
+    |> @tracer.close_span()
+
+    increment_job_stop_counter(job, state)
+
+    add_job_duration_value(job, duration, state)
+  end
+
+  def oban_job_exception(_event, %{duration: duration}, metadata, _config) do
+    span =
+      @tracer.current_span()
+      |> @span.set_attribute("state", to_string(metadata[:state]))
+      |> @span.set_attribute("worker", to_string(metadata[:job].worker))
+
+    if Map.has_key?(metadata, :kind) and Map.has_key?(metadata, :reason) and
+         Map.has_key?(metadata, :stacktrace) do
+      @span.add_error(span, metadata[:kind], metadata[:reason], metadata[:stacktrace])
+    end
+
+    @tracer.close_span(span)
+
+    increment_job_stop_counter(metadata[:job], metadata[:state])
+
+    add_job_duration_value(metadata[:job], duration, metadata[:state])
+  end
+
+  def oban_insert_job_start(_event, _measurements, metadata, _config) do
+    span = @tracer.create_span("oban", @tracer.current_span)
+
+    @span.set_attribute(span, "appsignal:category", "insert_job.oban")
+
+    case metadata[:changeset] do
+      %{changes: %{worker: worker}} ->
+        @span.set_name(span, "Insert job (#{worker})")
+
+      _ ->
+        @span.set_name(span, "Insert job")
+    end
+  end
+
+  def oban_insert_job_stop(_event, _measurements, metadata, _config) do
+    case metadata[:changeset] do
+      %{changes: %{worker: worker}} ->
+        @appsignal.increment_counter("oban_job_insert", 1, %{worker: to_string(worker)})
+
+      _ ->
+        nil
+    end
+
+    @tracer.close_span(@tracer.current_span())
+  end
+
+  def oban_insert_job_exception(_event, _measurements, _metadata, _config) do
+    @tracer.close_span(@tracer.current_span())
+  end
+
+  defp increment_job_start_counter(job) do
+    tag_combinations = [
+      %{worker: to_string(job.worker), queue: to_string(job.queue)},
+      %{worker: to_string(job.worker)},
+      %{queue: to_string(job.queue)}
+    ]
+
+    Enum.each(tag_combinations, fn tags ->
+      @appsignal.increment_counter("oban_job_start", 1, tags)
+    end)
+  end
+
+  defp increment_job_stop_counter(job, state) do
+    tag_combinations = [
+      %{worker: to_string(job.worker), state: to_string(state)},
+      %{state: to_string(state)}
+    ]
+
+    Enum.each(tag_combinations, fn tags ->
+      @appsignal.increment_counter("oban_job_stop", 1, tags)
+    end)
+  end
+
+  defp add_job_duration_value(job, duration, state) do
+    tag_combinations = [
+      %{worker: to_string(job.worker), state: to_string(state)},
+      %{worker: to_string(job.worker)}
+    ]
+
+    Enum.each(tag_combinations, fn tags ->
+      @appsignal.add_distribution_value(
+        "oban_job_duration",
+        System.convert_time_unit(duration, :native, :millisecond),
+        tags
+      )
+    end)
+  end
+
+  defp add_job_queue_time_value(job) do
+    delay = DateTime.diff(job.attempted_at, job.scheduled_at, :millisecond)
+
+    @appsignal.add_distribution_value("oban_job_queue_time", delay, %{
```

Comment:
Is it common for Oban to run on different or many hosts? If so, it would be interesting to add the host as a tag for this metric. That way it becomes more visible which hosts have the longest queue delay. Tracking the queue latency would be better, but it probably requires a minutely probe as you described and we do for Sidekiq in Ruby.

Reply:
I think it's possible to run Oban on multiple hosts. However, I don't know if the queue delay (the delta time between when the job is scheduled to be ran, and when the job is _actually_ ran by a worker) would vary between hosts.

Running the scenario in my head, multiple hosts connected to the same queue should have similar queue times. Say, 10 jobs enter the queue now to be executed immediately, host A can run 3 jobs per minute, host B is "broken" and can run only 1 job per minute. Two minutes from now, host A would pick up its 7th job and host B would pick up its 3rd job -- both of which would register queue delays of two minutes.

Maybe adding the hostname would be good for _other_ metrics, like job duration or the job stop counter?

Reply:
> I don't know if the queue delay (the delta time between when the job is scheduled to be ran, and when the job is actually ran by a worker) would vary between hosts.

I suppose it's more of a problem per queue. If you have split up queues per host. If y you can see this metric per queue (like it does not), you can figure out which host is a problem, if there are multiple, if necessary.

Let's leave it for now then.

> Maybe adding the hostname would be good for other metrics, like job duration or the job stop counter?

Job duration maybe, job count no. I don't care which worker processes more jobs. It's interesting data to see if some hosts are processing slower or have a higher queue to catch up to.

Reply:
> It's interesting data to see if some hosts [...] have a higher queue to catch up to.

Maybe I'm not understanding what you mean, but I don't think that's how multiple host setups work with Oban. You can have several hosts running workers from the same queue, but the queue is still one single queue in Postgres, not one per host -- meaning the "slow" host and the "fast" host have the same amount of jobs to catch up to.

[If you want it to work that way, you can configure a queue per host, tell each host to only listen to its queue, and round-robin message insertion between the queues somehow -- but it's not the default behaviour]

Reply:
Yep, you're right. I think halfway through I mixed it up with job duration, not queue duration.

Reply:
Added hostname for job duration -- we can reevaluate later.